Sources/GRPC/Compression/Zlib.swift

```diff
@@ -0,0 +1,422 @@
+    }
+  }
+
+  // MARK: Inflate (decompression)
+
+  /// Creates a new decompressor for the given compression format.
+  ///
+  /// This decompressor is only suitable for decompressing whole messages at a time. Callers
+  /// must `reset()` the decompressor between subsequent calls to `inflate`.
+  ///
+  /// - Parameter format:The expected compression type.
+  class Inflate {
+    private var stream: ZStream
+    private let format: CompressionFormat
+
+    init(format: CompressionFormat) {
+      self.stream = ZStream()
+      self.format = format
+      self.initialize()
+    }
+
+    deinit {
+      self.end()
+    }
+
+    /// Resets decompression state. This must be called after each call to `inflate` if more
+    /// messages are to be decompressed by this instance.
+    func reset() {
+      let rc = CGRPCZlib_inflateReset(&self.stream.zstream)
+
+      // Possible return codes:
+      // - Z_OK
+      // - Z_STREAM_ERROR: the source stream state was inconsistent.
+      //
+      // If we're in an inconsistent state we can just replace the stream and initialize it.
+      switch rc {
+      case Z_OK:
+        ()
+
+      case Z_STREAM_ERROR:
+        self.end()
+        self.stream = ZStream()
+        self.initialize()
+
+      default:
+        preconditionFailure("inflateReset: unexpected return code rc=\(rc)")
+      }
+    }
+
+    /// Inflate the readable bytes from the `input` buffer into the `output` buffer.
+    ///
+    /// - Parameters:
+    ///   - input: The buffer read compressed bytes from.
+    ///   - output: The buffer to write the decompressed bytes into.
+    /// - Returns: The number of bytes written into `output`.
+    @discardableResult
+    func inflate(_ input: inout ByteBuffer, into output: inout ByteBuffer) throws -> Int {
+      return try input.readWithUnsafeMutableReadableBytes { inputPointer -> (Int, Int) in
+        // Setup the input buffer.
+        self.stream.availableInputBytes = inputPointer.count
+        self.stream.nextInputBuffer = CGRPCZlib_castVoidToBytefPointer(inputPointer.baseAddress!)
+
+        defer {
+          self.stream.availableInputBytes = 0
+          self.stream.nextInputBuffer = nil
+        }
+
+        // We don't know how large the output will be; we'll try 2x the input size.
+        var outBufferSize = inputPointer.count * 2
+
+        var inflated = false
+        var bytesWritten = 0
+
+        while !inflated {
+          bytesWritten = try output.writeWithUnsafeMutableBytes(minimumWritableBytes: outBufferSize) { outputPointer in
+            let inflateResult = try self.stream.inflate(
+              outputBuffer: CGRPCZlib_castVoidToBytefPointer(outputPointer.baseAddress!),
+              outputBufferSize: outputPointer.count
+            )
+
+            switch inflateResult.outcome {
+            case .complete:
+              inflated = true
+            case .outputBufferTooSmall:
+              outBufferSize *= 2
```

Comment:
Could we do better than this with an educated guess? Can't Zlib tell you how large the decompressed data is by e.g. reading the compression header?

Reply:
I couldn't find anything in the docs about this and there's nothing equivalent to `deflateBound` for `inflate`, it seems. In the gRPC lib they allocate 1KB slices until they have enough space but their API is better suited to dealing with slices anyway.

We could do something similar; decompress into multiple buffers, find how much capacity we need for one, and then allocate enough to write them all into a single buffer. That way we'd only allocate ~2x the total output size. 

I think we'd need to profile this some more first though. IIRC zlib suggests that the compression ratio for many use cases is between 2 and 5 so it's not clear cut to me that allocating multiple buffers would actually be better.

If you're okay with this for the time being I'll add a comment to explain this.

Reply:
This is completely fine then; not worth more effort. I was just hoping for an easy win ;-)